crates/bevy_transform/src/transform_propagate_system.rs

```diff
@@ -1,61 +1,46 @@
 use crate::components::*;
 use bevy_ecs::prelude::*;
+use bevy_math::Mat4;
 
+// ANSWERME: make this take all transforms and perform in recursion (looking at compute shader)
 pub fn transform_propagate_system(
-    mut root_query: Query<
-        Without<Parent, (Option<&Children>, &mut Transform, Option<&LocalTransform>)>,
-    >,
-    mut local_transform_query: Query<(&mut Transform, &LocalTransform, Option<&Children>)>,
+    mut root_query: Query<Without<Parent, (Option<&Children>, &mut Transform)>>,
+    mut transform_query: Query<(&mut Transform, Option<&Children>)>,
 ) {
-    for (children, mut transform, local_transform) in &mut root_query.iter() {
-        if let Some(local_transform) = local_transform {
-            transform.value = local_transform.0;
-        }
+    for (children, mut transform) in &mut root_query.iter() {
+        transform.apply_parent_matrix(None);
 
         if let Some(children) = children {
             for child in children.0.iter() {
-                propagate_recursive(*transform, &mut local_transform_query, *child);
+                propagate_recursive(*transform.global_matrix(), &mut transform_query, *child);
             }
         }
     }
 }
 
+// ANSWERME: maybe speed this up with compute
 fn propagate_recursive(
-    parent_local_to_world: Transform,
-    local_transform_query: &mut Query<(&mut Transform, &LocalTransform, Option<&Children>)>,
+    parent: Mat4,
+    transform_query: &mut Query<(&mut Transform, Option<&Children>)>,
     entity: Entity,
 ) {
     log::trace!("Updating Transform for {:?}", entity);
-    let local_transform = {
-        if let Ok(local_transform) = local_transform_query.get::<LocalTransform>(entity) {
-            *local_transform
-        } else {
-            log::warn!(
-                "Entity {:?} is a child in the hierarchy but does not have a LocalTransform",
-                entity
-            );
-            return;
-        }
-    };
 
-    let new_transform = Transform {
-        value: parent_local_to_world.value * local_transform.0,
-        sync: true,
-    };
+    let global_matrix = {
+        let mut transform = transform_query.get_mut::<Transform>(entity).unwrap();
```

Comment:
we can resolve the `correct_children` unit test failure by returning if the entity does not have a transform or it does not exist.

```rust
let global_matrix = {
    if let Ok(mut transform) = transform_query.get_mut::<Transform>(entity) {
        transform.apply_parent_matrix(Some(parent));
        *transform.global_matrix()
    } else {
        return
    }
};
```

That being said, I do think the current failure exposes a (pre-existing) bug in our hierarchy maintenance code:

The children array isn't cleaned up when entities are despawned. This is something we should fix in a separate pr.

Reply:
So should I just paste the code you provided, or should we leave it in?

Reply:
Yeah just paste that code in for now. I'll create an issue to follow up.